kitsune/users/api.py

```diff
@@ -206,60 +205,54 @@ def get_last_answer_date(self, profile):
-
-    def validate_username(self, attrs, source):
-        obj = self.object
-        if obj is None:
-            # This is a create
-            if User.objects.filter(username=attrs['user.username']).exists():
-                raise ValidationError('A user with that username exists')
+    def validate(self, data):
+        if data.get('name') is None:
+            username = data.get('user', {}).get('username')
+            data['name'] = username
+
+        return data
+
+    def create(self, validated_data):
+        user_data = validated_data.pop('user')
+        u = RegistrationProfile.objects.create_inactive_user(
+            user_data['username'],
+            user_data['password'],
+            user_data['email'])
+        p = u.profile
+        for key, val in validated_data.items():
+            setattr(p, key, val)
+        p.save()
+        return p
+
+    def update(self, instance, validated_data):
+        if 'user' in validated_data:
+            user_data = validated_data.pop('user')
+            for key, val in user_data.items():
+                setattr(instance.user, key, val)
+            instance.user.save()
+        return super(ProfileSerializer, self).update(instance, validated_data)
+
+    def validate_username(self, username):
+        if re.match(r'^[\w.-]{4,30}$', username) is None:
+            raise ValidationError('Usernames may only be letters, numbers, "." and "-".')
+
+        if self.instance:
+            # update
+            if username != self.instance.user.username:
+                raise ValidationError({'username': "Can't change this field."})
         else:
-            # This is an update
-            new_username = attrs.get('user.username', obj.user.username)
-            if new_username != obj.user.username:
-                raise ValidationError("Can't change this field.")
+            # create
+            if User.objects.filter(username=username).exists():
+                raise ValidationError({'username': 'A user with that username exists'})
 
-        if re.match(r'^[\w.-]{4,30}$', attrs['user.username']) is None:
-            raise ValidationError(
-                'Usernames may only be letters, numbers, "." and "-".')
+        return username
 
-        return attrs
-
-    def validate_display_name(self, attrs, source):
-        if attrs.get('name') is None:
-            attrs['name'] = attrs.get('user.username')
-        return attrs
-
-    def validate_email(self, attrs, source):
-        email = attrs.get('user.email')
-        if email and User.objects.filter(email=email).exists():
-            raise ValidationError('A user with that email address '
-                                  'already exists.')
-        return attrs
+    def validate_email(self, email):
+        if not self.instance:
+            # create
+            if User.objects.filter(email=email).exists():
+                raise ValidationError({'email': 'A user with that email address already exists.'})
```

Comment:
Where this is specifically validating the email address, does it need to take a dict arg or is a string arg ok?